src/state_machine/coordinator/fire.rs

```diff
@@ -943,91 +943,101 @@ impl<Aggregator: AggregatorTrait> Coordinator<Aggregator> {
+            return Ok(());
+        };
 
-            // check that the signer_id exists in the config
-            let signer_public_keys = &self.config.signer_public_keys;
-            if !signer_public_keys.contains_key(&sig_share_response.signer_id) {
-                warn!(signer_id = %sig_share_response.signer_id, "No public key in config");
-                return Ok(());
-            };
+        let response_info = self.message_nonces.entry(self.message.clone()).or_default();
+        if response_info
+            .sign_wait_signer_ids
+            .contains(&sig_share_response.signer_id)
+        {
+            response_info
+                .sign_wait_signer_ids
+                .remove(&sig_share_response.signer_id);
+        } else {
+            warn!(
+                "Sign round {} SignatureShareResponse for round {} from signer {} not in the wait list",
+                self.current_sign_id, sig_share_response.sign_id, sig_share_response.signer_id,
+            );
+            return Ok(());
+        }
 
-            // check that the key_ids match the config
-            let Some(signer_key_ids) = self
-                .config
-                .signer_key_ids
-                .get(&sig_share_response.signer_id)
-            else {
-                warn!(signer_id = %sig_share_response.signer_id, "No keys IDs configured");
-                return Ok(());
-            };
+        debug!(
+            "Sign round {} SignatureShareResponse for round {} from signer {} ({}/{} key_ids). Waiting on {:?}",
+            self.current_sign_id,
+            sig_share_response.sign_id,
+            sig_share_response.signer_id,
+            response_info.sign_recv_key_ids.len(),
+            response_info.nonce_recv_key_ids.len(),
+            response_info.sign_wait_signer_ids
+        );
+
+        if sig_share_response.dkg_id != self.current_dkg_id {
+            return Err(Error::BadDkgId(
+                sig_share_response.dkg_id,
+                self.current_dkg_id,
+            ));
+        }
+        if sig_share_response.sign_id != self.current_sign_id {
+            return Err(Error::BadSignId(
+                sig_share_response.sign_id,
+                self.current_sign_id,
+            ));
+        }
 
-            let mut sig_share_response_key_ids = HashSet::new();
-            for sig_share in &sig_share_response.signature_shares {
-                for key_id in &sig_share.key_ids {
-                    sig_share_response_key_ids.insert(*key_id);
-                }
-            }
+        // check that the signer_id exists in the config
+        let signer_public_keys = &self.config.signer_public_keys;
+        if !signer_public_keys.contains_key(&sig_share_response.signer_id) {
+            warn!(signer_id = %sig_share_response.signer_id, "No public key in config");
+            return Err(Error::MissingPublicKeyForSigner(
+                sig_share_response.signer_id,
+            ));
```

Comment:
sanity check: this and the other new errors are "safe" since if we are in `gather_sig_shares` we already committed to a signer set, so if we are missing details about them we would not be able to succeed this round, no matter what. Right?